dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/message/ProgramMessageController.java

```diff
@@ -0,0 +1,125 @@
+import org.hisp.dhis.render.RenderService;
+import org.hisp.dhis.security.RequiresAuthority;
+import org.hisp.dhis.webapi.controller.AbstractFullReadOnlyController;
+import org.hisp.dhis.webapi.mvc.annotation.ApiVersion;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.RestController;
+
+/** Zubair <rajazubair.asghar@gmail.com> */
+@RestController
+@RequestMapping("/api/messages")
+@RequiredArgsConstructor
+@ApiVersion({DhisApiVersion.DEFAULT, DhisApiVersion.ALL})
+public class ProgramMessageController extends AbstractFullReadOnlyController<ProgramMessage> {
+  private final ProgramMessageService programMessageService;
+  private final ProgramMessageRequestParamsMapper paramsMapper;
+  private final RenderService renderService;
+
+  @RequiresAuthority(anyOf = F_MOBILE_SENDSMS)
+  @GetMapping(produces = APPLICATION_JSON_VALUE)
+  @ResponseBody
+  public ResponseEntity<List<ProgramMessage>> getProgramMessages(
+      ProgramMessageRequestParams requestParams) throws BadRequestException, ConflictException {
+    ProgramMessageQueryParams params = paramsMapper.ValidateAndMap(requestParams);
+
+    return ResponseEntity.ok(programMessageService.getProgramMessages(params));
+  }
+
+  @RequiresAuthority(anyOf = F_MOBILE_SENDSMS)
+  @GetMapping(value = "/scheduled/sent", produces = APPLICATION_JSON_VALUE)
+  @ResponseBody
+  public List<ProgramMessage> getScheduledSentMessage(ProgramMessageRequestParams requestParams)
+      throws BadRequestException, ConflictException {
+    requestParams.setMessageStatus(ProgramMessageStatus.SENT);
+    ProgramMessageQueryParams params = paramsMapper.ValidateAndMap(requestParams);
+
+    return programMessageService.getProgramMessages(params);
+  }
+
+  @RequiresAuthority(anyOf = {F_MOBILE_SENDSMS, F_SEND_EMAIL})
+  @PostMapping(consumes = APPLICATION_JSON_VALUE, produces = APPLICATION_JSON_VALUE)
+  @ResponseBody
+  public BatchResponseStatus sendMessages(HttpServletRequest request) throws IOException {
+    ProgramMessagesImportBatch batch =
+        renderService.fromJson(request.getInputStream(), ProgramMessagesImportBatch.class);
+
+    for (ProgramMessage programMessage : batch.getProgramMessages()) {
+      programMessageService.validatePayload(programMessage);
+    }
+
+    return programMessageService.sendMessages(batch.getProgramMessages());
+  }
+
+  @RequiresAuthority(anyOf = F_MOBILE_SENDSMS)
+  @PutMapping(value = "/{uid}", produces = APPLICATION_JSON_VALUE)
+  public WebMessage updateProgramMessage(
```

Comment:
Before we were allowing changing every single field of a `ProgramMessage`, now just the `status`, are we sure nobody was updating other fields?

Anyway, we should either maintain the other endpoint and deprecate it or agree on introducing a breaking change

Reply:
yes, ProgramMessage is an internal object created by `notificationServiceand`  and used by `messageSender`.
I hardly see any use case of changing `PrograMessage` fields other than `status`.